# Homework 1 - Introduction and first look at the data
## Deadline: September 21st, at noon (12:00)

# 1. Administrative requirements (0.5 points)

Please read carefully the following rules and instructions.

## Homework rules

The following rules apply on all 10 homeworks within this course, unless stated otherwise later.

#### No publishing or sharing of solutions

We kindly ask you not to publish or share the solutions to this or following homeworks (not even after the deadline, as some parts might be reused in the coming years).

#### Solving individually, no electronic sharing

The homeworks must be solved individually. You may ask or provide help to each other in trying to understand the tasks, but tasks must be solved and written down individually. Making any single line of code or text from your solution electronically available to other students is forbidden. If the code that you have written has a bug and you cannot find it yourself, it is fine to ask help from a more experienced programmer without providing him/her electronic access to your code (screenshot image of a few lines of code is fine, but copy-pasted code is not fine).

#### Using standard Python packages

You are free to use standard Python packages in your code, unless the particular task is stating otherwise.

#### Limitations on using public code

We sometimes request to implement an algorithm by yourself without using machine learning libraries such as `sklearn`, because it helps you to understand and remember it.  You then have an option to use some code from public external sources (e.g. Stack Overflow) as part of your code or as source of inspiration, but you must then cite it in the homework and make sure you understand the code. You can receive up to 80% of the maximal points per each subtask where you used public external code. If you use but don't cite a piece of public external code then we consider it to be plagiarism. Use of external code that is not publicly available online is forbidden.

NB! As an exception to the above, for visualization purposes (plotting) you may use any public code without citation.

#### Plagiarism cases

Violation of the above rules is considered plagiarism. If you get caught with a clear case of plagiarism:

- If there are no previous problems, depending on the situation, you might get away with just getting 0 points for the task and get a warning from us. Note that the curriculum manager would also be informed.
- If the problem appears several times, or we know that you have already warnings from other courses, the dean's office will be informed, and this can lead to expelling.

## How to solve the homework?

#### Please no deletion of cells, only adding solutions

Please do not delete any of the cells that have been given to you in the homework. If you accidentally do so, please add the deleted content back (`Edit -> Undo Deleted Cell` might help). Please only edit the cells where you need to insert your solutions. It is fine to have your solution consisting of many cells, so feel free to add cells if needed. Please make sure that the solution to each subtask is in the dedicated slot, e.g. it is not sufficient to just state that you already answered this question in the previous subtask.

These formatting requests are to make our grading simpler. We are using automatic tools to rearrange solutions, e.g. collecting solutions of the same subtask by all students into a single notebook for faster manual grading. If you delete some cells of task description then you risk losing points, although we try to notice and grade manually all such cases.

#### No need for code comments (unless stated otherwise)

We have no capacity to grade the way that you write code. Thus it is not compulsory to write comments into your code, just write enough to facilitate your coding.

#### Please read the tasks and instructions carefully

There are different kinds of tasks in homeworks. Some require answering questions or discussing some topic, in this case, your text should be comprehensive and written with full sentences. Some subtasks can have multiple questions, in such case please make sure you address all questions.

#### Tracking your time

We will appreciate if you track your time spent on solving the homework and report it in the dedicated cells at the end of the homework. This is not compulsory and does not affect your grade in any way. The collected information will be used to improve future homeworks.

#### Before submission

Please make sure that you do `Kernel -> Restart and Run All` on your Jupyter notebook before submission. This is something that we will do anyways before starting to grade, and by doing it yourself you can ensure that everything works as intended. Please make sure that your code does not throw errors half way through, because then whatever comes after that in the notebook would not be run. If you are not able to fully solve some coding task then at least please leave it at a state where it does not throw errors, commenting out some lines if necessary.

#### Submit only .ipynb file

Please submit only the .ipynb Jupyter notebook file, unless stated otherwise. We will run the notebook with all the required files (e.g. CSV files with data) in the same folder.

#### Deadline

The homework deadline is a hard deadline, meaning that you will get zero points for the homework even if you are late for just 1 minute. In case of medical reasons for being late please bring a certificate from the doctor and you can still get full points.

#### Questions?

Please ask on the forum if you have any questions regarding the above instructions or if you have problems understanding the tasks. We are happy to try to help :-)


## Homework format, discussion and feedback

Please read carefully the following instructions about homework format, discussion and feedback 

https://courses.cs.ut.ee/2020/ids/fall/Main/HomeworkFormat

**<font color='red'>Answer:</font>** The most important sentece for me is "You will not have to be correct or be able to answer everything. You just have to try and think along. Don't be scared to answer incorrectly or be wrong." 

# 2. Jupyter Notebook tutorial video (0.5 points)

First, please run the following cell by clicking inside it and typing "ctrl+enter", watch the following video "Jupyter Notebook Tutorial" that will pop, and answer all the questions below. Please answer the questions by replacing "Answer goes here" with your answer. Also, we recommend you to follow the examples in the video and try it out by yourself.

In [1]:
%%HTML 
<iframe width="560" height="315" src="https://www.youtube.com/embed/HW29067qVWk" frameborder="0" allowfullscreen></iframe>

 **a. What does the number inside "In [?]" mean before each cell?**

**<font color='red'>Answer:</font>** The "ln[?]" is corresponding to the number of cells you executed before this cell. For example, if you have not executed any cells, the first one you will execute will be the "ln[1]", and it doesn't matter if it is the first cell in the document.
If you run another cell, and then get back to this first cell, it will be now "ln[3]".

**b. What are the advantages (or disadvantages) of using Jupyter notebook compared to other programming environments (feel free to compare against any other environments that you have used in the past)?**

**<font color='red'>Answer:</font>** There is a lot of advantages of using Jupyter notebook, compared to other programming environments. 
- It is open-source, and free (and this, for a programmer, it's the best)
- You can execute cell one by one, so you can easily test your programm
- In Data Science, Jupyter's DataFrame are really simple and easy to read
- You can switch between programming cells and text cells, in order to make a really understandable document
- Finally, it is also easier to share code using Jupyter

Of course there is also some disadvantages, like :
- If there is too muche code, Jupyter have is own limit
- Sometimes, if we're not used to, it can be confusing if we are executing cells in a wrong order

# 3. Get started with Jupyter Notebook and pandas package  (1 point)

In this course, we are going to use pandas data structures and data analysis tool. Please make sure that you have gone through the tutorials presented in the second practice session: https://courses.cs.ut.ee/2020/ids/fall/Main/PracticeSessions -> Tutorial for DataFrames and Plotting. After that please answer the following questions below about the abalone dataset. The abalone dataset is included in the IDS2019_HW01.zip container. The same dataset can also be downloaded from __[here as CSV](https://courses.cs.ut.ee/2017/DM/fall/uploads/Main/abalone.csv)__ . Short description of variables is __[here as TXT](https://courses.cs.ut.ee/2017/DM/fall/uploads/Main/abalone_desc_2.txt)__ .

First, let's import pandas package as pd and read data from the CSV file into a pandas DataFrame.

In [4]:
import pandas as pd
data = pd.read_csv("abalone.csv")

Question 0 is an example of how you should answer the following questions. First, in the #TODO cell you have to write the code representing your solution. And in the answer cell write the answer with your own words. It does not have to be a full sentence.

 **0. What is overall memory consumption of the data?** (EXAMPLE)

In [5]:
# TODO - here you have to write the code how did you find the answer.
data.memory_usage(deep=True).sum()

102128

**<font color='red'>Answer:</font>** The total memory usage is 98128 bytes (98,128 kb).

 **a. What are the column names of the dataset?** 

In [7]:
# TODO
data.columns

Index(['Gender', 'Length', 'Diameter', 'Height', 'Weight', 'Rings'], dtype='object')

**<font color='red'>Answer:</font>** The column names are :
- Gender
- Length
- Diameter
- Height
- Weight
- Rings

**b. How many observations (i.e. rows) are in this data frame?**

In [5]:
# TODO
data.shape

(1000, 6)

**<font color='red'>Answer:</font>** In this data frame, there is 1000 rows and 6 columns.

**c. Print the first 5 lines from the dataset. What are the values of feature "Rings" of the printed observations?**

In [6]:
# TODO
data.head()

Gender  Length  Diameter  Height  Weight  Rings
0      F   0.505     0.385   0.135  0.6185     12
1      F   0.650     0.475   0.165  1.3875      9
2      I   0.520     0.380   0.135  0.5395      8
3      F   0.550     0.425   0.140  0.9520      7
4      M   0.640     0.525   0.180  1.3135     10

**<font color='red'>Answer:</font>** The values of feature "Rings" are [12,9,8,7,10] for the 5 first lines.

**d. Extract the last 3 rows of the data frame. What is the "Weight" of these abalones?**

In [7]:
# TODO
data.tail(3)

Gender  Length  Diameter  Height  Weight  Rings
997      M   0.660     0.500   0.165  1.3195      9
998      I   0.525     0.400   0.130  0.6455      8
999      M   0.515     0.395   0.135  1.0070      8

**<font color='red'>Answer:</font>** The value of "Weight" in the 3 last rows are [1.3195,0.6455,1.0070]

**e. What is the value of diameter in the row 577?**

In [8]:
# TODO
#data.head()
data.iloc[576, 2]

0.51

**<font color='red'>Answer:</font>** The value of diameter in the row 577 is 0.42 (we type 578 because the first line is 0).

**f. What is the mean of the height column?**<br>
Hint 1: Use pandas describe() method

In [9]:
# TODO
data.describe()
#data.mean()

Length     Diameter       Height       Weight        Rings
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000
mean      0.522760     0.405955     0.141100     0.825540    11.318000
std       0.120064     0.098832     0.057064     0.490380    47.227697
min       0.075000     0.055000     0.000000     0.002000     1.000000
25%       0.450000     0.345000     0.115000     0.438875     8.000000
50%       0.545000     0.420000     0.145000     0.801000     9.000000
75%       0.615000     0.480000     0.165000     1.146250    11.000000
max       0.815000     0.650000     1.130000     2.555000  1500.000000

**<font color='red'>Answer:</font>** The mean of the column "Height" is 0.141100

**g. Extract the subset of rows of the data frame where gender is M and weight values are below 0.75. What is the mean of diameter in this subset?**

In [10]:
# TODO
data[(data["Gender"] == "M") & (data["Weight"] <= 0.75)].Diameter.mean()

0.3426470588235295

**<font color='red'>Answer:</font>** The diameter's mean where gender is M and weight values are below 0.75 is : 0.3426470588235295

**h. What is the minimum of length when rings is equal to 18?**

In [11]:
# TODO
#data[(data["Rings"] == 18)].head()
data[(data["Rings"] == 18)].Length.min()

0.465

**<font color='red'>Answer:</font>** The minimum of lenght when rings is equal 18 is 0.465

**i. Is the weight of abalones related to how many rings they have? Please provide evidence and explanatory text with the conclusion?** <br>Hint 1:use groupby and mean()

In [9]:
# TODO
data.groupby("Rings").mean()["Weight"]

Rings
1       0.002000
3       0.024500
4       0.055036
5       0.132586
6       0.266913
7       0.416463
8       0.633463
9       0.849236
10      1.018775
11      1.145120
12      1.051500
13      1.098090
14      1.261604
15      1.073864
16      1.164417
17      1.274455
18      1.076625
19      1.094750
20      1.102278
21      1.676000
22      1.157000
23      1.043250
27      1.212500
1500    0.802500
Name: Weight, dtype: float64

**<font color='red'>Answer:</font>** No, rings and weight are not related. If we have 10 rings for a weight's mean of 1.01, we should have for 23 rings a weight's mean of (1.01*23)/10 = 2,32. But, we have for 23 rings a weight's mean equals to 1.04.
So, we can conlude that rings and weight are not related.

# 4. TED talks  (1 point)

Look at https://www.springboard.com/blog/11-must-watch-ted-talks-on-data-science/. Please watch at least 3 of these talks. For each of those three talks answer the following 2 questions (altogether we expect 3x2=6 answers). <b>Please don’t forget to include the titles of the videos</b> that you watched in your answers. Full score will be given only for comprehensive answers. For example, the answers "I liked the talk" and "I liked the talk because it was about data science" are not good enough.

### How We Found the Worst Place to Park in New York City — Using Big Data

**a. Was it interesting to you and what did you like or didn’t like about this?**

**<font color='red'>Answer:</font>** I like it because he talk about the necessity of sharing data, and the power that every citizen have. In the video, because he did some research, he found out a problem in NYC and NYC listen to him. So, maybe not every problem but a lot of problems (especially small ones) who can be solved by research and juste a little action. I also found interesting the fact that we should be able to access data much easier than pdf.

**b. What was the key message that you would tell about to your friend?**

**<font color='red'>Answer:</font>** The key message is "Every citizen can share is data, in order to make everyone life better". You don't have to be part of a tech cumminuty, you can just share and/or see any data and the government needs to make an easy access to these data.

### How I Hacked Online Dating

**a. Was it interesting to you and what did you like or didn’t like about this?**

**<font color='red'>Answer:</font>** I like a lot the presentation, because it was catchy and really fun. I found really interesting that she made through something we all know, a market research and all the data analysis. She proved that everything could be analyze with data, even dating.

**b. What was the key message that you would tell about to your friend?**

**<font color='red'>Answer:</font>** The key message is "Each field, each area can be studied in order to improve this field, by analyzing data. Of course data analysis is working a lot in business, but in fact, we can use it everywhere."

### Your Social Media “likes” Expose More than You Think

**a. Was it interesting to you and what did you like or didn’t like about this?**

**<font color='red'>Answer:</font>** I was interested on this TedTalk because I'm really concerned about privacy (I'm not even on Facebook, and I know you share a lot of data even if you are not on any social media). The way she talked about privacy, and the fact that it would be great if she's not able to "steal" data from people, I think it is very honest. Also, the curly fries's like is a really good example of how data scientist measure your profile.

**b. What was the key message that you would tell about to your friend?**

**<font color='red'>Answer:</font>** The key message is "Just being on internet make you give a lot of information about yourself. The problem isn't using social medias but how social medias steal your data, and your not aware of this. So maybe you can be more carefull about what you like, because now you know that it will be used by some data scientist !"

# 5. Data for your life  (1 point)

Please think about how you could gather or find data and use or analyse it to improve something in your everyday life, such as related to home, friends, travel, school, work, etc. Alternatively, this idea could improve the life of someone you know personally. Please describe your idea and its potential impact in 100-300 words.

**<font color='red'>Answer:</font>** I'm not a big gamer, but with the two weeks quarantine, I restart to play a little bit in video games, essentially to CS:GO. To resume really quickly, it's a game where 2 teams of 5 are against :
- One team is the "terrorist" team : to win, they have to kill all the "anti-terrorist" team OR plant a bomb and protect her during a certain time
- The other team is the "anti-terrorist" team : to win, they have to kill all the "terrorist" team OR defuse the bomb.

First of all, why it will be useful. This videogame is very popular, and there is professional teams and world championship, where more and more of money is engaged. So, if there is money, I think some people should be interested into a Data Science project on this game.

Then, it's a videogame, so maybe collecting data can be really easy, because everything is already in a computer. There is also a lot of games around the world, so if we could extract all these data, we can have a large data base. We can track players around the map, famous location, weapons used, time, level of players, ...

Finally, all this study can improve a lot the way to play this game, in order to make some strategies. Maybe if we exploit data, we can see a strategy that is never used by professional teams. Also, like it is a videogame, there is maybe some "bug" in it that we can exploit ?

The difficulty for this project will maybe be that there is a lot of data, and it will be hard to found a right correlation between all of these.

 ## <font color='red'>This was the last task! Please restart the kernel and run all before submission! (`Kernel -> Restart and Run All`)</font>

## How long did it take you to solve the homework?

Please answer as precisely as you can. It does not affect your points or grade in any way. It is okey, if it took 0.5 hours or 24 hours. Please count in astronomical hours (1 hour = 60 minutes) and not academic hours (1 hour = 45 minutes). The collected information will be used to improve future homeworks.
<font color = "red"><b> Please replace "X" with the amount of hours and don't change the unit "hours" or write any other text. </b></font>

**<font color='red'>Task 1 (please change X in the next cell into your estimate)</font>**

0.1 hour

**<font color='red'>Task 2 (please change X in the next cell into your estimate)</font>**

0.5 hour

**<font color='red'>Task 3 (please change X in the next cell into your estimate)</font>**

1 hour

**<font color='red'>Task 4 (please change X in the next cell into your estimate)</font>**

1 hour

**<font color='red'>Task 5 (please change X in the next cell into your estimate)</font>**

0.1 hour

**<font color='red'>TOTAL (please change X in the next cell into your estimate)</font>**

2.7 hours

**<font color='red'>THANK YOU FOR YOUR EFFORT!</font>**